triggers/service/src/main/scala/com/digitalasset/daml/lf/engine/trigger/ServiceMain.scala

```diff
@@ -86,34 +89,53 @@ class Server(dar: Option[Dar[(PackageId, Package)]], jdbcConfig: Option[JdbcConf
         // will still call addPackage even if we already fed the
         // package via the callback but this is harmless and not
         // expensive.
+        @scala.annotation.tailrec
         def go(r: Result[Unit]): Unit = r match {
           case ResultDone(()) => ()
           case ResultNeedPackage(pkgId, resume) =>
             go(resume(darMap.get(pkgId)))
           case _ => throw new RuntimeException(s"Unexpected engine result $r")
         }
+
         go(compiledPackages.addPackage(pkgId, pkg))
     }
   }
 
   private def getRunningTrigger(uuid: UUID): RunningTrigger = {
-    triggers.get(uuid).get // TODO: Improve as might throw NoSuchElementException.
+    triggers(uuid) // TODO: Improve as might throw NoSuchElementException.
   }
 
-  private def addRunningTrigger(t: RunningTrigger): Unit = {
-    triggers = triggers + (t.triggerInstance -> t)
-    triggersByToken = triggersByToken + (t.jwt -> (triggersByToken.getOrElse(t.jwt, Set()) + t.triggerInstance))
+  private def addRunningTrigger(t: RunningTrigger): Either[String, Unit] = {
+    triggerDao match {
```

Comment:
If you don’t want to go down the route of using sqlite with an in-memory backend so you always have a DB, I think it would be useful to have an abstraction layer somewhere. Inlining this every time we write to the DB is going to become a mess and it’s super easy to have the in-memory mode diverge from the DB-backed mode.

Reply:
Yes, I'll refactor the in-memory logic into the TriggerDao in a future PR (either next or after `stop` is implemented).